storage/aptosdb/src/pruner/worker.rs

```diff
@@ -0,0 +1,140 @@
+            for db_pruner in &self.db_pruners {
+                let result = db_pruner
+                    .lock()
+                    .prune(Self::MAX_VERSIONS_TO_PRUNE_PER_BATCH);
+                match result {
+                    Ok(_) => {}
+                    Err(_) => {
+                        // On error, stop retrying vigorously by making next recv() blocking.
+                        self.blocking_recv = true;
+                    }
+                }
+            }
+            for db_pruner in &self.db_pruners {
+                // if any of the pruner has pending pruning, then we don't block on receive
+                if db_pruner.lock().is_pruning_pending() {
+                    self.blocking_recv = false
+                }
+            }
+            self.record_progress();
+        }
+    }
+
+    fn record_progress(&mut self) {
+        let mut updated_least_readable_versions: Vec<Version> = Vec::new();
+        for x in &self.db_pruners {
+            updated_least_readable_versions.push(x.lock().least_readable_version())
+        }
+        *self.least_readable_versions.lock() = updated_least_readable_versions;
+    }
+
+    /// Tries to receive all pending commands, blocking waits for the next command if no work needs
+    /// to be done, otherwise quits with `true` to allow the outer loop to do some work before
+    /// getting back here.
+    ///
+    /// Returns `false` if `Command::Quit` is received, to break the outer loop and let
+    /// `work_loop()` return.
+    fn receive_commands(&mut self) -> bool {
+        loop {
+            let command = if self.blocking_recv {
+                // Worker has nothing to do, blocking wait for the next command.
+                self.command_receiver
+                    .recv()
+                    .expect("Sender should not destruct prematurely.")
+            } else {
+                // Worker has pending work to do, non-blocking recv.
+                match self.command_receiver.try_recv() {
+                    Ok(command) => command,
+                    // Channel has drained, yield control to the outer loop.
+                    Err(_) => return true,
+                }
+            };
+
+            match command {
+                // On `Command::Quit` inform the outer loop to quit by returning `false`.
+                Command::Quit => return false,
+                Command::Prune { target_db_versions } => {
+                    for (new_target_version, pruner) in
+                        zip_eq(&target_db_versions, &self.db_pruners)
+                    {
+                        if *new_target_version > pruner.lock().target_version() {
+                            // Switch to non-blocking to allow some work to be done after the
+                            // channel has drained.
+                            self.blocking_recv = false;
+                        }
+                        pruner.lock().set_target_version(*new_target_version);
```

Comment:
Is it true that they always share the same target version? In that case shall we just make them share a `AtomicVersion`?

Reply:
target version will be different for different pruner - this depends on the pruning window size configured per pruner. Am I missing something?

Reply:
I see. There are actually two distinct target versions, and most of the pruners will share one of them. It's fine like this.